File: 0143-reorder-list/0143-reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        Stack<ListNode> stack = new Stack<>();

        ListNode ref1 = head; 
        ListNode ref2 = head;
        
        while (ref2.next != null) {
            ref1 = ref1.next;
            ref2 = ref2.next;
            if (ref2.next != null)
                ref2 = ref2.next;
        }
        
        while (ref1 != null) {
            stack.add(ref1);
            ref1 = ref1.next;
        }
        
        ListNode ref3 = head;
        while (!stack.isEmpty()) {
            ListNode temp = stack.pop();
            temp.next = ref3.next;
            ref3.next = temp;
            ref3 = temp.next;
        }
        ref3.next = null;
    }
}

File: 0143-reorder-list/0143-reorder-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        # find the mid
        ref = mid = head
        while ref != None and ref.next != None:
            mid = mid.next
            ref = ref.next.next
        
        # reverse second half
        prev = None
        curr = mid
        while curr:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp
        
        # merge the first half and the second half
        ref1 = head
        while prev.next != None:
            temp = ref1.next
            temp2 = prev.next
            ref1.next = prev
            prev.next = temp
            ref1 = temp
            prev = temp2